lightning/src/ln/keysend_payments_tests.rs

```diff
@@ -0,0 +1,557 @@
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+	let payer_pubkey = nodes[0].node.get_our_node_id();
+	let payee_pubkey = nodes[1].node.get_our_node_id();
+
+	let _chan = create_chan_between_nodes(&nodes[0], &nodes[1]);
+	let route_params = RouteParameters::from_payment_params_and_value(
+		PaymentParameters::for_keysend(payee_pubkey, 40, false), 10_000);
+	let network_graph = nodes[0].network_graph;
+	let first_hops = nodes[0].node.list_usable_channels();
+	let scorer = test_utils::TestScorer::new();
+	let random_seed_bytes = chanmon_cfgs[1].keys_manager.get_secure_random_bytes();
+	let route = find_route(
+		&payer_pubkey, &route_params, &network_graph, Some(&first_hops.iter().collect::<Vec<_>>()),
+		nodes[0].logger, &scorer, &Default::default(), &random_seed_bytes
+	).unwrap();
+
+	let test_preimage = PaymentPreimage([42; 32]);
+	let test_secret = PaymentSecret([43; 32]);
+	let payment_hash = PaymentHash(Sha256::hash(&test_preimage.0).to_byte_array());
+	let session_privs = nodes[0].node.test_add_new_pending_payment(payment_hash,
+		RecipientOnionFields::secret_only(test_secret), PaymentId(payment_hash.0), &route).unwrap();
+	nodes[0].node.test_send_payment_internal(&route, payment_hash,
+		RecipientOnionFields::secret_only(test_secret), Some(test_preimage),
+		PaymentId(payment_hash.0), None, session_privs).unwrap();
+	check_added_monitors!(nodes[0], 1);
+
+	let updates = get_htlc_update_msgs!(nodes[0], nodes[1].node.get_our_node_id());
+	assert_eq!(updates.update_add_htlcs.len(), 1);
+	assert!(updates.update_fulfill_htlcs.is_empty());
+	assert!(updates.update_fail_htlcs.is_empty());
+	assert!(updates.update_fail_malformed_htlcs.is_empty());
+	assert!(updates.update_fee.is_none());
+	nodes[1].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &updates.update_add_htlcs[0]);
+
+	nodes[1].logger.assert_log_contains("lightning::ln::channelmanager", "We don't support MPP keysend payments", 1);
+}
+
+#[test]
+fn bad_inbound_payment_hash() {
+	// Add coverage for checking that a user-provided payment hash matches the payment secret.
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+	node_cfgs[0].keys_manager.get_inbound_payment_key_material();
+
+	let highest_seen_timestamp = bitcoin::blockdata::constants::genesis_block(bitcoin::Network::Testnet).header.time;
+	let node_signer = node_cfgs[0].keys_manager;
+	let inbound_pmt_key_material = node_signer.get_inbound_payment_key_material();
+	let expanded_inbound_key = inbound_payment::ExpandedKey::new(&inbound_pmt_key_material);
+
+	let (_, payment_hash, payment_secret) = get_payment_preimage_hash!(&nodes[0]);
+	let payment_data = msgs::FinalOnionHopData {
+		payment_secret,
+		total_msat: 100_000,
+	};
+
+
```

Comment:
I have refactored this test, can you go through this whether this behaves as expected ?

Reply:
Basically I tried to untie the dependency between the channelmanagers internal state, I dont know whether it is hacky or not